Bottom: 4ce01cb530e72185faaa7408d722fd3566f4bfbd
Top:    833ddabb12146cf8d987cebc154ab4699f783a21
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-02-28 16:05:33 +0800

Refresh of refresh-temp-0

---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 8f6d0bd..1cb8057 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -474,6 +474,26 @@ static unsigned int virtnet_get_headroom(struct virtnet_info *vi)
 	return vi->xdp_queue_pairs ? VIRTIO_XDP_HEADROOM : 0;
 }
 
+static struct sk_buff *virtnet_skb_xdp(struct receive_queue *rq,
+				       struct sk_buff *skb)
+{
+	struct bpf_prog *xdp_prog;
+	int ret;
+
+	rcu_read_lock();
+	xdp_prog = rcu_dereference(rq->xdp_prog);
+	if (xdp_prog) {
+		ret = do_xdp_generic(xdp_prog, skb);
+		if (ret != XDP_PASS) {
+			rcu_read_unlock();
+			return NULL;
+		}
+	}
+	rcu_read_unlock();
+
+	return skb;
+}
+
 static struct sk_buff *receive_small(struct net_device *dev,
 				     struct virtnet_info *vi,
 				     struct receive_queue *rq,
@@ -490,7 +510,6 @@ static struct sk_buff *receive_small(struct net_device *dev,
 			      SKB_DATA_ALIGN(sizeof(struct skb_shared_info));
 	struct page *page = virt_to_head_page(buf);
 	unsigned int delta = 0;
-	struct page *xdp_page;
 	bool sent, skb_xdp = false;
 	int err;
 
@@ -601,26 +620,6 @@ static struct sk_buff *receive_big(struct net_device *dev,
 	return NULL;
 }
 
-static struct sk_buff *virtnet_skb_xdp(struct receive_queue *rq,
-				       struct sk_buff *skb)
-{
-	struct bpf_prog *xdp_prog;
-	int ret;
-
-	rcu_read_lock();
-	xdp_prog = rcu_dereference(rq->xdp_prog);
-	if (xdp_prog) {
-		ret = do_xdp_generic(xdp_prog, skb);
-		if (ret != XDP_PASS) {
-			rcu_read_unlock();
-			return NULL;
-		}
-	}
-	rcu_read_unlock();
-
-	return skb;
-}
-
 static struct sk_buff *receive_mergeable(struct net_device *dev,
 					 struct virtnet_info *vi,
 					 struct receive_queue *rq,
